luarules/gadgets/unit_projectile_leash.lua

```diff
@@ -0,0 +1,163 @@
+local forcedDescentUpdateFrames = math.ceil(Game.gameSpeed / 5)
+local minimumThresholdRange = 10 --so that starburst missiles don't trigger edgyWatch during ascent
+local compoundingMultiplier = 1.1 --compounding multiplier that influences the arc at which projectiles are forced to descend
+local descentSpeedStartingMultiplier = 0.1
+
+--functions
+local spGetUnitPosition = Spring.GetUnitPosition
+local mathSqrt = math.sqrt
+local mathMax = math.max
+local spGetProjectilePosition = Spring.GetProjectilePosition
+local spSetProjectileCollision = Spring.SetProjectileCollision
+local spGetProjectileVelocity = Spring.GetProjectileVelocity
+local spSetProjectileVelocity = Spring.SetProjectileVelocity
+
+--tables
+local defWatchTable = {}
+local lazyProjectileWatch = {}
+local edgyProjectileWatch = {}
+local forcedDescentTable = {}
+
+
+for weaponDefID, weaponDef in pairs(WeaponDefs) do
+	if weaponDef.customParams.projectile_leash_range then
+		defWatchTable[weaponDefID] = {}
+		defWatchTable[weaponDefID].leashRange = tonumber(weaponDef.customParams.projectile_leash_range)
+	end
+	if weaponDef.customParams.projectile_leash_range or weaponDef.customParams.projectile_overrange_distance then
+		defWatchTable[weaponDefID] = defWatchTable[weaponDefID] or {}
+		defWatchTable[weaponDefID].range = weaponDef.range
+		defWatchTable[weaponDefID].weaponRange = weaponDef.range
+		defWatchTable[weaponDefID].overRange = tonumber(weaponDef.customParams.projectile_overrange_distance) or weaponDef.range
+		defWatchTable[weaponDefID].rangeThreshold = math.max((defWatchTable[weaponDefID].overRange - weaponDef.projectilespeed * lazyUpdateFrames), minimumThresholdRange)
+		defWatchTable[weaponDefID].weaponDefID = weaponDefID
+
+		local destructionMethod = weaponDef.customParams.projectile_destruction_method or "explode"
+		if destructionMethod == "descend" then
+			defWatchTable[weaponDefID].descentMethod = true
+		else
+			defWatchTable[weaponDefID].explodeMethod = true
+		end
+
+		Script.SetWatchWeapon(weaponDefID, true)
+	end
+end
+
+
+local function projectileOverRangeCheck(proOwnerID, weaponRange,  leashRange, originX, originZ, projectileX, projectileZ)
+	local dx1 = originX - projectileX
+	local dz1 = originZ - projectileZ
+	local distanceToOrigin = mathSqrt(dx1 * dx1 + dz1 * dz1)
+
+	if distanceToOrigin > weaponRange then
+		if leashRange then
+			local distanceToOwner
+			local ownerX, ownerY, ownerZ = spGetUnitPosition(proOwnerID)
+			if ownerX then
+				local dx2 = ownerX - projectileX
+				local dz2 = ownerZ - projectileZ
+				distanceToOwner = mathSqrt(dx2 * dx2 + dz2 * dz2)
+				if distanceToOwner > leashRange then
+					return true
+				end
+			end
+			return false
+		end
+		return true
+	end
+	return false
+end
+
+local function projectileIsCloseToEdge(rangeThreshold, originX, originZ, projectileX, projectileZ)
+	local dx1 = originX - projectileX
+	local dz1 = originZ - projectileZ
+	local distanceToOrigin = mathSqrt(dx1 * dx1 + dz1 * dz1)
```

Comment:
Same comment here re: Sqrt and storing rangeThresholdSq instead of rangeThreshold.

Reply:
Yep!